File: Motocitizen/src/motocitizen/ui/activity/MainScreenActivity.java
package motocitizen.ui.activity;

import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import kotlin.Unit;
import motocitizen.MyApp;
import motocitizen.content.Content;
import motocitizen.content.accident.Accident;
import motocitizen.datasources.network.ApiResponse;
import motocitizen.datasources.preferences.Preferences;
import motocitizen.geo.geolocation.MyLocationManager;
import motocitizen.geo.maps.MainMapManager;
import motocitizen.main.R;
import motocitizen.permissions.Permissions;
import motocitizen.router.Router;
import motocitizen.ui.changelog.ChangeLog;
import motocitizen.ui.rows.accident.AccidentRowFactory;
import motocitizen.ui.rows.accident.Row;
import motocitizen.ui.views.BounceScrollView;
import motocitizen.user.User;
import motocitizen.utils.Utils;

public class MainScreenActivity extends AppCompatActivity {
    private static final byte LIST = 0;
    private static final byte MAP  = 1;

    private ViewGroup   mapContainer;
    private ImageButton createAccButton;
    private ImageButton toAccListButton;
    private ImageButton toMapButton;
    private View        accListView;
    private ProgressBar progressBar;
    private ViewGroup   listContent;

    private MenuItem       refreshItem;
    private MainMapManager map;
    private boolean inTransaction = false;
    private byte    currentScreen = LIST;
    private static ActionBar actionBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_screen_activity);
        actionBar = getSupportActionBar();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (map == null) map = new MainMapManager(this);
        wakeUpLocationUpdate();
        showChangeLogIfUpdated();

        disableDialOnTablets();

        bindViews();
        setUpListeners();

        setPermissions();
        showCurrentFrame();
        redraw();
        getAccidents();
    }

    private void wakeUpLocationUpdate() {
        Permissions.INSTANCE.requestLocation(this,
                                             () -> {
                                                 MyLocationManager.INSTANCE.wakeup();
                                                 map.enableLocation();
                                                 return Unit.INSTANCE;
                                             },
                                             () -> Unit.INSTANCE);
    }

    private void disableDialOnTablets() {
        if (!getPackageManager().hasSystemFeature(PackageManager.FEATURE_TELEPHONY)) {
            this.findViewById(R.id.dial_button).setEnabled(false);
        }
    }

    private void showChangeLogIfUpdated() {
        if (!Preferences.INSTANCE.getNewVersion()) return;
        AlertDialog changeLogDlg = ChangeLog.INSTANCE.getDialog(this);
        changeLogDlg.show();
        Preferences.INSTANCE.setNewVersion(false);
    }

    private void setUpListeners() {
        createAccButton.setOnClickListener(v -> Router.INSTANCE.goTo(this, Router.Target.CREATE));
        toAccListButton.setOnClickListener(v -> showListFrame());
        toMapButton.setOnClickListener(v -> showMapFrame());
        this.findViewById(R.id.dial_button).setOnClickListener(v -> Router.INSTANCE.dial(this, getString(R.string.phone)));
        ((BounceScrollView) this.findViewById(R.id.accListRefresh)).setOverScrollListener(this::getAccidents);
    }

    private void bindViews() {
        accListView = this.findViewById(R.id.acc_list);
        progressBar = (ProgressBar) this.findViewById(R.id.progressBar);
        mapContainer = (ViewGroup) this.findViewById(R.id.google_map);
        createAccButton = (ImageButton) this.findViewById(R.id.add_point_button);
        toAccListButton = (ImageButton) this.findViewById(R.id.list_button);
        toMapButton = (ImageButton) this.findViewById(R.id.map_button);
        listContent = (ViewGroup) this.findViewById(R.id.accListContent);
    }

    @Override
    protected void onPause() {
        super.onPause();
        Permissions.INSTANCE.requestLocation(this,
                                             () -> {MyLocationManager.INSTANCE.sleep(); return Unit.INSTANCE;},
                                             () -> Unit.INSTANCE);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (intent.hasExtra("toMap")) {
            //mainFragment.toMap(intent.getExtras().getInt("toMap", 0));
            toMap(intent.getExtras().getInt("toMap", 0));
            intent.removeExtra("toMap");
        }
        if (intent.hasExtra("toDetails")) {
            intent.removeExtra("toDetails");
        }
        setIntent(intent);
    }

    //todo exterminatus static
    public static void updateStatusBar(String address) {
        String subTitle = "";
        //Делим примерно пополам, учитывая пробел или запятую
        int commaPos = address.lastIndexOf(",", address.length() / 2);
        int spacePos = address.lastIndexOf(" ", address.length() / 2);

        if (commaPos != -1 || spacePos != -1) {
            subTitle = address.substring(Math.max(commaPos, spacePos) + 1);
            address = address.substring(0, Math.max(commaPos, spacePos));
        }

        actionBar.setTitle(address);
        if (!subTitle.isEmpty()) actionBar.setSubtitle(subTitle);
    }

    public void setPermissions() {
        createAccButton.setVisibility(User.INSTANCE.isStandard() ? View.VISIBLE : View.INVISIBLE);
    }

    public void redraw() {
        List<Row> newList = new ArrayList();
        for (Accident accident : Content.INSTANCE.getVisibleReversed()) {
            newList.add(AccidentRowFactory.INSTANCE.make(this, accident));
        }
        listContent.removeAllViews();

        for (Row row : newList) {
            listContent.addView(row);
        }
        map.placeAccidents(this);
    }

    private Unit getAccidents() {
        if (inTransaction) return Unit.INSTANCE;
        if (MyApp.isOnline(this)) {
            startRefreshAnimation();
            Content.INSTANCE.requestUpdate(this::updateCompleteCallback);
        } else {
            Toast.makeText(this, getString(R.string.inet_not_available), Toast.LENGTH_LONG).show();
        }
        return Unit.INSTANCE;
    }

    private Unit updateCompleteCallback(ApiResponse stub) {
        runOnUiThread(() -> {
            stopRefreshAnimation();
            redraw();
        });
        return Unit.INSTANCE;
    }

    private void setRefreshAnimation(boolean status) {
        progressBar.setVisibility(status ? View.VISIBLE : View.INVISIBLE);
        inTransaction = status;
        //TODO костыль
        if (refreshItem != null) refreshItem.setVisible(!status);
    }

    private void stopRefreshAnimation() {
        setRefreshAnimation(false);
    }

    private void startRefreshAnimation() {
        setRefreshAnimation(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.small_settings_menu, menu);
        refreshItem = menu.findItem(R.id.action_refresh);
        if (inTransaction) refreshItem.setVisible(false);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.small_menu_refresh:
                getAccidents();
                return true;
            case R.id.small_menu_settings:
                Router.INSTANCE.goTo(this, Router.Target.SETTINGS);
                return true;
            case R.id.small_menu_about:
                Router.INSTANCE.goTo(this, Router.Target.ABOUT);
                return true;
            case R.id.action_refresh:
                getAccidents();
                return true;
            case R.id.do_not_disturb:
                item.setIcon(Preferences.INSTANCE.getDoNotDisturb() ? R.drawable.ic_lock_ringer_on_alpha : R.drawable.ic_lock_ringer_off_alpha);
                Preferences.INSTANCE.setDoNotDisturb(!Preferences.INSTANCE.getDoNotDisturb());
                return true;
        }
        return false;
    }

    private void setFrame(byte target) {
        currentScreen = target;
        toAccListButton.setAlpha(target == LIST ? 1f : 0.3f);
        toMapButton.setAlpha(target == MAP ? 1f : 0.3f);
        accListView.animate().translationX(target == LIST ? 0 : -Utils.getWidth(this) * 2);
        mapContainer.animate().translationX(target == MAP ? 0 : Utils.getWidth(this) * 2);
    }

    private void showListFrame() {
        setFrame(LIST);
    }

    private void showMapFrame() {
        setFrame(MAP);
    }

    private void showCurrentFrame() {
        setFrame(currentScreen);
    }

    public void toMap(int id) {
        showMapFrame();
        map.jumpToPoint(Content.INSTANCE.accident(id).getCoordinates());
    }
}

File: Motocitizen/src/motocitizen/ui/activity/MainScreenActivity.kt
package motocitizen.ui.activity

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ProgressBar
import android.widget.Toast
import com.google.android.gms.maps.model.LatLng
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.runBlocking
import motocitizen.MyApp
import motocitizen.content.Content
import motocitizen.datasources.preferences.Preferences
import motocitizen.geo.geolocation.MyLocationManager
import motocitizen.geo.maps.MainMapManager
import motocitizen.main.R
import motocitizen.permissions.Permissions
import motocitizen.router.Router
import motocitizen.ui.changelog.ChangeLog
import motocitizen.ui.rows.accident.AccidentRowFactory
import motocitizen.ui.views.BounceScrollView
import motocitizen.user.User
import motocitizen.utils.bindView
import motocitizen.utils.displayWidth

class MainScreenActivity : AppCompatActivity() {
    companion object {
        private const val LIST: Byte = 0
        private const val MAP: Byte = 1
    }

    private val mapContainer: ViewGroup by bindView(R.id.google_map)
    private val createAccButton: ImageButton by bindView(R.id.add_point_button)
    private val toAccListButton: ImageButton by bindView(R.id.list_button)
    private val toMapButton: ImageButton by bindView(R.id.map_button)
    private val accListView: View by bindView(R.id.acc_list)
    private val progressBar: ProgressBar by bindView(R.id.progressBar)
    private val listContent: ViewGroup by bindView(R.id.accListContent)

    private var refreshItem: MenuItem? = null
    private lateinit var map: MainMapManager
    private var inTransaction = false
    private var currentScreen = LIST

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_screen_activity)
        map = MainMapManager(this)
    }

    override fun onResume() {
        super.onResume()
        wakeUpLocationUpdate()
        showChangeLogIfUpdated()
        MyLocationManager.subscribeToLocationUpdate("MAIN", this::updateStatusBar)
        disableDialOnTablets()

        setUpListeners()

        setPermissions()
        showCurrentFrame()
        redraw()
        accidents
    }

    private fun wakeUpLocationUpdate() {
        Permissions.requestLocation(this, {
            MyLocationManager.wakeup()
        }) { }
    }

    private fun disableDialOnTablets() {
        if (packageManager.hasSystemFeature(PackageManager.FEATURE_TELEPHONY)) return
        findViewById(R.id.dial_button).isEnabled = false
    }

    private fun showChangeLogIfUpdated() {
        if (!Preferences.newVersion) return
        val changeLogDlg = ChangeLog.getDialog(this)
        changeLogDlg.show()
        Preferences.newVersion = false
    }

    private fun setUpListeners() {
        createAccButton.setOnClickListener { Router.goTo(this, Router.Target.CREATE) }
        toAccListButton.setOnClickListener { showListFrame() }
        toMapButton.setOnClickListener { showMapFrame() }
        findViewById(R.id.dial_button).setOnClickListener { Router.dial(this, getString(R.string.phone)) }
        (findViewById(R.id.accListRefresh) as BounceScrollView).setOverScrollListener { accidents }
    }

    override fun onPause() {
        super.onPause()
        MyLocationManager.unSubscribe("MAIN")
        Permissions.requestLocation(this, { MyLocationManager.sleep() }) { }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (intent.hasExtra("toMap")) {
            //mainFragment.toMap(intent.getExtras().getInt("toMap", 0));
            toMap(intent.extras.getInt("toMap", 0))
            intent.removeExtra("toMap")
        }
        if (intent.hasExtra("toDetails")) {
            intent.removeExtra("toDetails")
        }
        setIntent(intent)
    }

    private fun setPermissions() {
        createAccButton.visibility = if (User.isStandard) View.VISIBLE else View.INVISIBLE
    }

    private fun redraw() {
        val newList = runBlocking {
            Content.getVisibleReversed()
                    .map { async(CommonPool) { AccidentRowFactory.make(this@MainScreenActivity, it) } }
                    .map { it.await() }
        }
        listContent.removeAllViews()

        newList.forEach(listContent::addView)

        map.update()
    }

    //todo WTF!?
    private val accidents: Unit
        get() {
            if (inTransaction) return
            if (MyApp.isOnline(this)) {
                startRefreshAnimation()
                Content.requestUpdate { updateCompleteCallback() }
            } else {
                Toast.makeText(this, getString(R.string.inet_not_available), Toast.LENGTH_LONG).show()
            }
        }

    private fun updateCompleteCallback() {
        runOnUiThread {
            stopRefreshAnimation()
            redraw()
        }
    }

    //todo extract progressBar to separate class
    private fun stopRefreshAnimation() {
        setRefreshAnimation(false)
    }

    private fun startRefreshAnimation() {
        setRefreshAnimation(true)
    }

    private fun setRefreshAnimation(status: Boolean) {
        progressBar.visibility = if (status) View.VISIBLE else View.INVISIBLE
        inTransaction = status
        //TODO костыль
        if (refreshItem != null) refreshItem!!.isVisible = !status
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.small_settings_menu, menu)
        refreshItem = menu.findItem(R.id.action_refresh)
        if (inTransaction) refreshItem!!.isVisible = false

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.small_menu_refresh  -> accidents
            R.id.small_menu_settings -> Router.goTo(this, Router.Target.SETTINGS)
            R.id.small_menu_about    -> Router.goTo(this, Router.Target.ABOUT)
            R.id.action_refresh      -> accidents
            R.id.do_not_disturb      -> {
                item.setIcon(if (Preferences.doNotDisturb) R.drawable.ic_lock_ringer_on_alpha else R.drawable.ic_lock_ringer_off_alpha)
                Preferences.doNotDisturb = !Preferences.doNotDisturb
            }
            else                     -> return false
        }
        return true
    }

    private fun showListFrame() {
        setFrame(LIST)
    }

    private fun showMapFrame() {
        setFrame(MAP)
    }

    private fun showCurrentFrame() {
        setFrame(currentScreen)
    }

    private fun setFrame(target: Byte) {
        currentScreen = target
        toAccListButton.alpha = if (target == LIST) 1f else 0.3f
        toMapButton.alpha = if (target == MAP) 1f else 0.3f
        accListView.animate().translationX((if (target == LIST) 0 else -displayWidth() * 2).toFloat())
        mapContainer.animate().translationX((if (target == MAP) 0 else displayWidth() * 2).toFloat())
    }

    private fun toMap(id: Int) {
        showMapFrame()
        map.centerOnAccident(Content[id]!!)
    }

    //todo refactor
    private fun updateStatusBar(latLng: LatLng) {
        var address = MyLocationManager.getAddress(latLng)
        var subTitle = ""
        //Делим примерно пополам, учитывая пробел или запятую
        val commaPos = address.lastIndexOf(",", address.length / 2)
        val spacePos = address.lastIndexOf(" ", address.length / 2)

        if (commaPos != -1 || spacePos != -1) {
            subTitle = address.substring(Math.max(commaPos, spacePos) + 1)
            address = address.substring(0, Math.max(commaPos, spacePos))
        }

        actionBar?.title = address
        if (!subTitle.isEmpty()) actionBar?.subtitle = subTitle
    }
}
